### PR TITLE
When the update key is changed, subscribe to the new key.

### PR DESCRIPTION
**Pull request by ArneBab**

### CODE REVIEW

File: src/freenet/node/updater/NodeUpdater.java [hunk 1]
```diff
@@ -511,8 +511,14 @@ public synchronized boolean canUpdateNow() {
 	/** Called when the fetch URI has changed. No major locks are held by caller. 
 	 * @param uri The new URI. */
 	public void onChangeURI(FreenetURI uri) {
-		kill();
+		kill(); // unsubscribes from the old uri
 		this.URI = uri;
+		try {
```
Comment (Bombe) [inline src/freenet/node/updater/NodeUpdater.java:516]: This looks suspiciously like the `start()` method. Maybe we should simply call that method here?

**Review by Bombe — CHANGES_REQUESTED** [review]

File: src/freenet/node/updater/NodeUpdater.java (hunk 1 above)
Comment (ArneBab) [reply src/freenet/node/updater/NodeUpdater.java]: That’s a good point. It’s not exactly the same (we won’t want to stop updating when the key is invalid), but very close. I factored out a subscribe() method.

File: src/freenet/node/updater/NodeUpdater.java [hunk 2]
```diff
@@ -91,15 +91,23 @@ public abstract class NodeUpdater implements ClientGetCallback, USKCallback, Req
 	}
 
 	void start() {
+        try {
+            this.subscribe();
```
Comment (Bombe) [inline src/freenet/node/updater/NodeUpdater.java:95]: The `this` qualifier is over-doing it a bit, methinks. 😄

File: src/freenet/node/updater/NodeUpdater.java [hunk 3]
```diff
@@ -91,15 +91,23 @@ public abstract class NodeUpdater implements ClientGetCallback, USKCallback, Req
...
+
+    void subscribe() throws MalformedURLException {
 		try {
 			// because of UoM, this version is actually worth having as well
 			USK myUsk = USK.create(URI.setSuggestedEdition(currentVersion));
 			core.getUskManager().subscribe(myUsk, this, true, getRequestClient());
 		} catch(MalformedURLException e) {
 			Logger.error(this, "The auto-update URI isn't valid and can't be used");
-			manager.blow("The auto-update URI isn't valid and can't be used", true);
+            throw e;
```
Comment (Bombe) [inline src/freenet/node/updater/NodeUpdater.java:108]: This irks me… catching an exception, just to log it, then re-throw it and ignore it in one of the methods. That feels like a lot of work.

Hmm, oh boy, here I go, making it overly complicated again… but what about:

```
private void subscribe(Runnable onError) {
    try {
    } catch (MalformedURLException e) {
        Logger.error(this, "The auto-update URI isn't valid and can't be used");
        onError.run();
    }
}
```
so we can have
```
void start() {
    subscribe(() -> manager.blow("The auto-update URI isn't valid and can't be used", true));
}
```
and
```
public void onChangeURI(FreenetURI uri) {
    kill();
    this.URI = uri;
    subscribe(() -> {});
    maybeUpdate();
}
```
?

**Review by Bombe — CHANGES_REQUESTED** [review]

File: src/freenet/node/updater/NodeUpdater.java [hunk 4]
```diff
@@ -91,15 +91,23 @@ public abstract class NodeUpdater implements ClientGetCallback, USKCallback, Req
...
 
 	void start() {
+        try {
+            this.subscribe();
+		} catch(MalformedURLException e) {
+            manager.blow("The auto-update URI isn't valid and can't be used", true);
+        }
+	}
+
+    void subscribe() throws MalformedURLException {
```
[the remark this answers is not in the document]
Comment (ArneBab) [reply src/freenet/node/updater/NodeUpdater.java]: Not a good reason, no — fixed. Thanks!

File: src/freenet/node/updater/NodeUpdater.java (hunk 2 above)
Comment (ArneBab) [reply src/freenet/node/updater/NodeUpdater.java]: Oh yes — I blame Javascript ☺

Thank you!

File: src/freenet/node/updater/NodeUpdater.java (hunk 3 above)
Comment (ArneBab) [reply src/freenet/node/updater/NodeUpdater.java]: I wanted to say “no, this makes it too non-obvious”, but it’s actually much nicer with the Runnable.

So: yes ☺

Thank you!

**Review by Bombe — APPROVED** [review]

**Comment by ArneBab** [comment]
merged — thank you for your review!

**Pull request merged**